solutions/usaco-623.mdx

```diff
@@ -2,115 +2,141 @@
 
-**Time Complexity:** $\mathcal{O}((N\text{max}(y_i))(\text{log}(\text{max}(y_i))+\text{log}(N)))$
+**Time Complexity:** $\mathcal{O}((N \cdot  \max \{ y_i = 1 .. n \}
+)(\text{log}(\max \{ y_i = 1 .. n \})+\text{log}(N)))$
 
 <LanguageSection>
 <CPPSection>
 
 ```cpp
 #include <bits/stdc++.h>
 using namespace std;
-typedef long long ll;
-typedef pair<int,int> pii;
-#define ff first
-#define ss second
-
-//BeginCodeSnip{Disjoint-Set Union}
-pii find(pii target, vector<vector<pii>>& top){
-	if (target == top[target.ff][target.ss]){
-		return target;
-	}
-	return target = find(top[target.ff][target.ss], top);
-}
 
-bool merge(pii a, pii b, vector<vector<pii>>& top, vector<vector<int>>& size){
-	a = find(a, top);
-	b = find(b, top);
-	if (a == b) {
-		return false;
-	}
-	if (size[a.ff][a.ss] < size[b.ff][b.ss]) {
-		swap(a, b);
+struct Edge {
+	int cost;
+	int i;
+	int j;
+};
+
+//BeginCodeSnip{DSU}
+struct DSU {
+	vector<int> e;
+	DSU(int N) { e = vector<int>(N, -1); }
+
+	// get representive component (uses path compression)
+	int get(int x) { return e[x] < 0 ? x : e[x] = get(e[x]); }
+
+	bool same_set(int a, int b) { return get(a) == get(b); }
+
+	int size(int x) { return -e[get(x)]; }
+
+	bool unite(int x, int y) {	// union by size
+		x = get(x), y = get(y);
+		if (x == y) return false;
+		if (e[x] > e[y]) swap(x, y);
+		e[x] += e[y];
+		e[y] = x;
+		return true;
 	}
-	top[b.ff][b.ss] = a;
-	size[a.ff][a.ss] += size[b.ff][b.ss];
-	return true;
-}
+};
 //EndCodeSnip
 
 int main() {
-	ifstream fin("fencedin.in");
+	freopen("fencedin.in", "r", stdin);
+	freopen("fencedin.out", "w", stdout);
+
+	int hdist, vdist, n, m;
+	cin >> hdist >> vdist >> n >> m;
+
+	vector<int> vx(n + 1), vy(m + 1);
 
-	// Sorting the fence lengths from shortest to longest
-	int xlimit, ylimit, n, m;
-	fin >> xlimit >> ylimit >> n >> m;
-	vector<int> vert(n + 1);
-	vector<int> horz(m + 1);
 	for (int i = 1; i <= n; i++) {
-		fin >> vert[i];
+		cin >> vx[i];
 	}
+
 	for (int i = 1; i <= m; i++) {
-		fin >> horz[i];
-	}
-	sort(vert.begin(), vert.end());
-	sort(horz.begin(), horz.end());
-	for (int i = 0; i < n; i++) {
-		vert[i] = vert[i + 1] - vert[i];
-	}
-	for (int i = 0; i < m; i++) {
-		horz[i] = horz[i + 1] - horz[i];
+		cin >> vy[i];
 	}
-	vert[n] = xlimit - vert[n];
-	horz[m] = ylimit - horz[m];
-	sort(vert.begin(), vert.end());
-	sort(horz.begin(), horz.end());
-
-	// Set up the DSU
-	vector<vector<int>> size(n + 1, vector<int>(m + 1, 1));
-	vector<vector<pii>> top(n + 1, vector<pii>(m + 1, make_pair(0, 0)));
-	for (int i = 0; i <= n; i++) {
-		for (int j = 0; j <= m; j++) {
-			top[i][j] = make_pair(i, j);
+
+	vx.push_back(hdist);
+	vy.push_back(vdist);
+
+	sort(vx.begin(), vx.end());
+	sort(vy.begin(), vy.end());
+
+	// add fences at borders to calculate edges at front/back
+	n += 2;
+	m += 2;
+
+	/*
+	 * we'll number sections from left
+	 * to right.
+	 *
+	 * -------------
+	 * |  1  |  2  |
+	 * |-----------|
+	 * |  3  |  4  |
+	 * -------------
+	 */
```

Comment:
```suggestion
	/*
	 * we'll number sections from left
	 * to right.
	 * 1 2
	 * 3 4
	 */
```

Reply:
what's wrong w this?

Reply:
feel like it makes more sense bc it shows the fences?

Reply:
takes too many lines imo

Reply:
eh, fair enough.